components/tidb_query_normal_expr/src/builtin_time.rs

```diff
@@ -547,6 +558,76 @@ impl ScalarFunc {
         Ok(None)
     }
 
+    #[inline]
+    pub fn sub_string_and_duration<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let arg0 = try_opt!(self.children[0].eval_string(ctx, row));
+        let arg1 = try_opt!(self.children[1].eval_duration(ctx, row));
+        if is_duration(std::str::from_utf8(&arg0)?) {
```

Comment:
Maybe we don't need this `is_duration`, we could just use the `Duration::parse`, if there is an `OK`, then the string must be in a valid duration format, otherwise, it is not.

Reply:
Oh, of course we can; I also considered this way at the first glance, at the same time I was worried that in this way `Duration::parse` would be executed twice each time (the first time to judge whether the string is valid duration, while the second time to parse). Maybe the performance would decrease?

Reply:
Or after I add the benchmark the performance can be compared.

Reply:
@andylokandy There has been a discussion on `is_duration` in this conversation; not sure whether it is better to just use `Duration::parse`. Maybe a decision can be made now?

This function is borrowed from TiDB.

BTW, before I considered to add a benchmark test, while I'm not familiar with adding benchmark in Rust.

Reply:
Tell me if I'm speaking stupid because I'm not familiar with the `AddStringAndDuration`: What is the difference between your implementation and directly `Duration::parse()` and then `Duration + Duration`?

Reply:
To be honest I'm not sure what's the benefit of the `is_duration` function, I just copied the implementation in TiDB. I'll fix it.

Reply:
done

Reply:
I think the whole PR can be rewritten by `Duration::parse` just like https://github.com/tikv/tikv/pull/8999

Reply:
done

Not sure, the implementation in TiDB checks 1) if the string can be parsed as a Duration, use `Duration::checked_add`; 2) otherwise try to use `DateTime::checked_add`. I'm curious about whether the 2) step can be skipped.

Reply:
@andylokandy @iosmanthus 

Sorry for ignoring an important background for `is_duration`:

`is_duration` cannot be replaced as `match Duration::parse`. When given an `DateTime` format string, `Duration::parse` can parse it without `Error` as well.

e.g. when given `2018-02-28 23:00:00`, the result of `Duration::parse` will be `Ok(Duration("23:00:00"))`. Then for `ADD_DATE("2018-02-28 23:00:00", "01:30:30.123456")`, the result will be `"24:30:30.123456"`, not `"2018-03-01 00:30:30.123456"` (the real result).

On the other hand, `DateTime::parse` cannot parse durations. e.g. `DateTime::parse("01:00:00.999999")` returns `Error(_)`.

Thus `is_duration`, `Duration::parse` & `DateTime::parse` ---- all of them shouldn't be skipped.

The commit will be reverted, and the code will be refactored as well. Sorry for ignoring these cases.

PTAL